--- paired-forecast-v2/ra4/namelist.input
+++ pristine-jan-ra4/case/namelist.input
@@ -62,10 +62,6 @@
  icloud                              = 1,
  cu_rad_feedback                     = .true.,.true.,.false.,
  num_land_cat                        = 24,
- rrtmgp_udm_frozen_optics             = 0,
- rrtmgp_data_path                    = '/NHNHOME/WORKSPACE/26weather002_A/yhlee/RRTMGP/build/udm-alternate-jan2000-data/paired-forecast-v2/ra4/rrtmgp_data',
- rrtmgp_ice_roughness                = 1,
- rrtmgp_udm_frozen_table             = '/NHNHOME/WORKSPACE/26weather002_A/yhlee/RRTMGP/build/udm-alternate-jan2000-data/paired-forecast-v2/ra4/frozen-ice-psd-moments.nc',
  sf_urban_physics                    = 0,     0,     0,
  fractional_seaice                   = 1,
  /
