--- source RA4 namelist (PR21 case)
+++ pristine RA4 namelist
@@ -63,7 +63,6 @@

  &physics
  use_mp_re                          = 1,
- rrtmgp_data_path                   = '/NHNHOME/WORKSPACE/26weather002_A/yhlee/RRTMGP/build/pr-wrf-rrtmgp/WRF/run',
  aer_opt                            = 0,
  cldovrlp                           = 2,
  swint_opt                          = 0,
@@ -89,8 +88,6 @@
  maxens2                             = 3,
  maxens3                             = 16,
  ensdim                              = 144,
- rrtmgp_udm_frozen_optics = 0,
- rrtmgp_udm_frozen_table = '/NHNHOME/WORKSPACE/26weather002_A/yhlee/RRTMGP/build/udm-frozen-runtime-wrf/source/validation/rrtmgp37/frozen-optics-lookup/nine-grid-o128-s50/frozen-ice-psd-moments.nc',
  /

  &dynamics
